cpp_generator.py

```diff
@@ -0,0 +1,367 @@
+
+    def __get_namespace_for_node(self, node: VSSNode) -> list[str]:
+        result = [n.name.lower().replace("switch", "switch_") for n in node.path][0:-1]
+        return result
+
+    def __generate_guard_name(self, node: VSSNode) -> str:
+        return "VMDL_" + "_".join([p.name.upper() for p in node.path]) + "_H"
+
+    def __gen_header(self, node: VSSNode):
+        guard_name = self.__generate_guard_name(node)
+        self.ctx_header.write(
+            f"""#ifndef {guard_name}
+            #define {guard_name}\n\n""",
+            strip_lines=True,
+        )
+
+        self.ctx_source.write(f"#include \"{node.name}.hpp\"\n\n")
+        self.ctx_source.write(self.__generate_opening_namespace_text("velocitas", node))
+
+    def __gen_footer(self, node: VSSNode):
+        self.ctx_header.write(self.__generate_closing_namespace_text("velocitas", node))
+        self.ctx_header.write(f"#endif // {self.__generate_guard_name(node)}\n")
+        self.ctx_source.write(self.__generate_closing_namespace_text("velocitas", node))
+
+    def __gen_imports(self, node: VSSNode):
+        self.ctx_header.write(f"#include \"sdk/DataPoint.h\"\n")
+        self.ctx_header.write(f"#include \"sdk/Model.h\"\n")
+        self.ctx_header.write("\n")
+        
+        for imp in sorted(self.includes):
+            if imp[0] == ".":
+                imp = imp[2:]
+            self.ctx_header.write(f"#include \"{imp}.hpp\"\n")
+
+        if len(self.includes) > 0:
+            self.ctx_header.write("\n")
+
+        for inc in sorted(self.external_includes):
+            self.ctx_header.write(f"#include <{inc}>\n")
+
+        if len(self.external_includes) > 0:
+            self.ctx_header.write("\n")
+
+        self.includes.clear()
+        self.external_includes.clear()
+
+    def __gen_model_docstring(self, node: VSSNode):
+        self.ctx_header.write(f'/** {node.name} model. */\n')
+
+    def __document_member(self, node: VSSNode):
+        self.ctx_header.write("/**\n")
+        if node.type.value == "attribute":
+            self.ctx_header.write(f"* {node.name}: {node.type.value} ({node.datatype.value})\n")
+        else:
+            self.ctx_header.write(f"* {node.name}: {node.type.value}\n")
+
+        self.ctx_header.write(f"* {node.description}\n")
+        self.ctx_header.write("*\n")
+        if len(node.comment) > 0:
+            self.ctx_header.write(f"* {node.comment}\n")
+            self.ctx_header.write("*\n")
+
+        if not isinstance(node.min, str) or not isinstance(node.max, str):
+            self.ctx_header.write(f"* Value range: [{node.min}, {node.max}]\n")
+        if hasattr(node, "unit"):
+            self.ctx_header.write(f"* Unit: {node.unit}\n")
+        if len(node.allowed) > 0:
+            allowed_values = ", ".join(node.allowed)
+            self.ctx_header.write(f"* Allowed values: {allowed_values}\n")
+        self.ctx_header.write("**/\n")
+
+    def __gen_collection_types(self, node: VSSNode, path: str) -> str:
```

Comment:
This method is quite long, can we split it?

Reply:
hmm yes, but lets discuss offline if this is better than